src/store/singing.ts

```diff
@@ -1011,62 +1011,142 @@ export const singingStore = createPartialStore<SingingStoreTypes>({
         return { engineId: state.engineId, styleId: state.styleId };
       };
 
-      const renderingRequested = () => state.renderingRequested;
-      const renderingStopRequested = () => state.renderingStopRequested;
+      // NOTE: 型推論でawaitの前か後かが考慮されないので、関数を介して取得する（型がbooleanになるようにする）
+      const startRenderingRequested = () => state.startRenderingRequested;
+      const stopRenderingRequested = () => state.stopRenderingRequested;
 
       const render = async () => {
-        if (!state.score) {
-          throw new Error("score is undefined.");
+        if (!state.score || !singChannel || !audioRenderer) {
+          throw new Error(
+            "score or singChannel or audioRenderer is undefined."
+          );
         }
+        const singChannelRef = singChannel;
+        const audioRendererRef = audioRenderer;
+
         // レンダリング中に変更される可能性のあるデータをコピーする
         const score = copyScore(state.score);
         const singer = getSinger();
 
-        await updatePhrases(singer, score);
-        if (renderingRequested() || renderingStopRequested()) {
+        preProcessing(score);
+
+        window.electron.logInfo("Updating phrases...");
+
+        const foundPhrases = await searchPhrases(singer, score);
```

Comment:
レンダリングが３段階になっているということがわかりやすいようにコメントがあると良さそうかもです。
こんな感じの理解です。

```suggestion
        // Score -> Phrase
        const foundPhrases = await searchPhrases(singer, score);
```